test/jdk/java/foreign/TestAddressHandle.java

```diff
@@ -0,0 +1,111 @@
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+/*
+ * @test
+ * @run testng/othervm -Djava.lang.invoke.VarHandle.VAR_HANDLE_GUARDS=true -Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=false -Xverify:all TestAddressHandle
+ * @run testng/othervm -Djava.lang.invoke.VarHandle.VAR_HANDLE_GUARDS=true -Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=true -Xverify:all TestAddressHandle
+ * @run testng/othervm -Djava.lang.invoke.VarHandle.VAR_HANDLE_GUARDS=false -Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=false -Xverify:all TestAddressHandle
+ * @run testng/othervm -Djava.lang.invoke.VarHandle.VAR_HANDLE_GUARDS=false -Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=true -Xverify:all TestAddressHandle
+ */
+
+import java.lang.invoke.*;
+import java.nio.ByteOrder;
+import jdk.incubator.foreign.*;
+
+import org.testng.annotations.*;
+import static org.testng.Assert.*;
+
+public class TestAddressHandle {
+    @Test(dataProvider = "addressHandles")
+    public void testAddressHandle(VarHandle addrHandle) {
+        VarHandle longHandle = MemoryHandles.varHandle(long.class, ByteOrder.nativeOrder());
+        try (MemorySegment segment = MemorySegment.allocateNative(8)) {
+            longHandle.set(segment.baseAddress(), 42L);
+            MemoryAddress address = (MemoryAddress)addrHandle.get(segment.baseAddress());
+            assertEquals(address.offset(), 42L);
+            try {
+                longHandle.get(address); // check OOB
+                fail();
+            } catch (UnsupportedOperationException ex) {
+                assertTrue(true);
+            }
+            addrHandle.set(segment.baseAddress(), address.addOffset(1));
+            long result = (long)longHandle.get(segment.baseAddress());
+            assertEquals(43L, result);
+        }
+    }
+
+    @Test(dataProvider = "addressHandles")
+    public void testNull(VarHandle addrHandle) {
+        VarHandle longHandle = MemoryHandles.varHandle(long.class, ByteOrder.nativeOrder());
+        try (MemorySegment segment = MemorySegment.allocateNative(8)) {
+            longHandle.set(segment.baseAddress(), 0L);
+            MemoryAddress address = (MemoryAddress)addrHandle.get(segment.baseAddress());
+            assertTrue(address == MemoryAddress.NULL);
+        }
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void testBadAdaptFloat() {
+        VarHandle floatHandle = MemoryHandles.varHandle(float.class, ByteOrder.nativeOrder());
+        MemoryHandles.asAddressVarHandle(floatHandle);
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void testBadAdaptDouble() {
+        VarHandle doubleHandle = MemoryHandles.varHandle(double.class, ByteOrder.nativeOrder());
+        MemoryHandles.asAddressVarHandle(doubleHandle);
+    }
```

Comment:
The adapter code also checks for `!carrier.isPrimitive()` that's also a case that could be tested.

Reply:
I'll add something